shell/platform/linux/fl_keyboard_manager.cc

```diff
@@ -0,0 +1,412 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/fl_keyboard_manager.h"
+
+#include <cinttypes>
+
+/* Declare and define FlKeyboardPendingEvent */
+
+/**
+ * FlKeyboardPendingEvent:
+ * A record for events that have been received by the manager, but
+ * dispatched to other objects, whose results have yet to return.
+ *
+ * This object is used by both the "pending_responds" list and the
+ * "pending_redispatches" list.
+ */
+G_DECLARE_FINAL_TYPE(FlKeyboardPendingEvent,
+                     fl_keyboard_pending_event,
+                     FL,
+                     KEYBOARD_PENDING_EVENT,
+                     GObject);
+
+struct _FlKeyboardPendingEvent {
+  GObject parent_instance;
+
+  // The target event.
+  //
+  // This is freed by #FlKeyboardPendingEvent.
+  FlKeyEvent* event;
+
+  // Self-incrementing ID attached to an event sent to the framework.
+  //
+  // Used to identify pending responds.
+  uint64_t sequence_id;
+  // The number of responders that haven't replied.
+  size_t unreplied;
+  // Whether any replied responders reported true (handled).
+  bool any_handled;
+
+  // A value calculated out of critical event information that can be used
+  // to identify redispatched events.
+  uint64_t hash;
+};
+
+G_DEFINE_TYPE(FlKeyboardPendingEvent, fl_keyboard_pending_event, G_TYPE_OBJECT)
+
+static void fl_keyboard_pending_event_dispose(GObject* object) {
+  // Redundant, but added so that we don't get a warning about unused function
+  // for FL_IS_KEYBOARD_PENDING_EVENT.
+  g_return_if_fail(FL_IS_KEYBOARD_PENDING_EVENT(object));
+
+  FlKeyboardPendingEvent* self = FL_KEYBOARD_PENDING_EVENT(object);
+  fl_key_event_dispose(self->event);
+  G_OBJECT_CLASS(fl_keyboard_pending_event_parent_class)->dispose(object);
+}
+
+static void fl_keyboard_pending_event_class_init(
+    FlKeyboardPendingEventClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_keyboard_pending_event_dispose;
+}
+
+static void fl_keyboard_pending_event_init(FlKeyboardPendingEvent* self) {}
+
+// Calculates a unique ID for a given FlKeyEvent object to use for
+// identification of responses from the framework.
+static uint64_t fl_keyboard_manager_get_event_hash(FlKeyEvent* event) {
+  // Combine the event timestamp, the type of event, and the hardware keycode
+  // (scan code) of the event to come up with a unique id for this event that
+  // can be derived solely from the event data itself, so that we can identify
+  // whether or not we have seen this event already.
+  return (event->time & 0xffffffff) |
+         (static_cast<uint64_t>(event->is_press ? GDK_KEY_PRESS
+                                                : GDK_KEY_RELEASE) &
+          0xffff)
+             << 32 |
```

Comment:
This has some weird indenting.  Maybe it would make sense to split into some const values and | those together for readability? (I know this is just copied from the previous implementation, but it's ugly. :-))

Reply:
Sure.

Speaking of this, this logic is a bit weird to me. I think this hash should _not_ include time, and _should_ include keyval. What was your idea?

Reply:
The idea was to identify the exact event from GTK, so the time is crucial to that. If you didn't include time, then a rapidly pressed key might mean that it matches a different down event for that key, and handle them out of order. These three attributes were enough to make the hash be unique for the fields we cared about.  You could add in keyval, or even substitute it for keycode if you want: I don't think that would make much difference.